### PR TITLE
User roles added

### PR DESCRIPTION
**Pull request by behica**
Updated app to use Pundit and added authorizations. I also updated the validation to `before_action` on `application_controller` to allow guests to browse wikis and read them, but require sign-in before creating new or editing. I also updated the `destroy?` method to authorize if user is admin.

The user story didn't make any mention of assigning wikis to users yet, so I left it out. Is this something I should have done with this section or should I do it once the app is complete?

### CODE REVIEW

File: app/policies/application_policy.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  def show?
+    scope.where(:id => record.id).exists?
+  end
+
+  def create?
+    false
+  end
+
+  def new?
+    create?
```
Comment (bobbilee19) [inline app/policies/application_policy.rb]: yep see just like this is what I was talking about!

File: app/policies/wiki_policy.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  def update?
+    user.present?
+  end
+  
+  def new?
+    user.present?
+  end
+  
+  def create?
+    user.present?
```
Comment (bobbilee19) [inline app/policies/wiki_policy.rb]: follow what you did in the last one above :)

File: bin/setup [hunk 3]
```diff
@@ -1,29 +1,34 @@
...
 require 'pathname'
+require 'fileutils'
+include FileUtils
 
 # path to your application root.
-APP_ROOT = Pathname.new File.expand_path('../../',  __FILE__)
+APP_ROOT = Pathname.new File.expand_path('../../', __FILE__)
 
-Dir.chdir APP_ROOT do
+def system!(*args)
```
Comment (bobbilee19) [inline bin/setup:9]: explain this method in your own words

File: db/schema.rb [hunk 4]
```diff
@@ -29,21 +28,20 @@
...
-  add_index "users", ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true
-
   create_table "wikis", force: :cascade do |t|
     t.string   "title"
     t.text     "body"
     t.boolean  "private"
     t.integer  "user_id"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.index ["user_id"], name: "index_wikis_on_user_id"
```
Comment (bobbilee19) [inline db/schema.rb:44]: does this make sense, why would we do it on a belongs_to relationship, reference the article i sent you

File: bin/setup (hunk 3 above)
Comment (behica) [reply bin/setup:9]: The system method is being passed an argument that can be multiple arguments. Those arguments are then put into a hash for the method to call on. The ! says this is a dangerous method that has the ability to change objects, or fail with exceptions.

The method itself is expecting that `system` be passed multiple arguments, or it aborts the environment load stating that the given arguments have failed.

File: app/policies/application_policy.rb (hunk 1 above)
Comment (behica) [reply app/policies/application_policy.rb]: This was the default given policy. I'll read up on why the methods where just `create?` is used in `def new?` and change my file accordingly. Thanks!

File: db/schema.rb (hunk 4 above)
Comment (behica) [reply db/schema.rb:44]: Yeah definitely. The article helps, and just using the reminder that the index is given here for a table reference helped re-organize my perspective. When I first saw index, I kept thinking 'why am I creating index views for my users on my wiki model?' and got stuck in that loop. Once you mentioned the table index the lighbulb came back on.

File: app/policies/wiki_policy.rb (hunk 2 above)
Comment (behica) [reply app/policies/wiki_policy.rb]: Let me know if I get this right...

My WikiPolicy inherits the default application policy, right? So in essence, I only need to add the authorizations I want changed to my wiki policy. For example, application policy says:
```
def destroy?
  false
end
```
Is this essentially saying that users are not allowed to call the `destroy` method? So in my wiki policy, I leave the `user.admin?` method for `destroy` and when it checks the wiki controller for the `def destroy` method and sees `authorize @wiki`, it checks to see if the user is an admin and either allows or rejects the method call.

In the end, I should only change / update my wiki policy where it doesn't agree with the default application policy, is that correct?

Comment (bobbilee19) [reply app/policies/wiki_policy.rb]: yes sir :)

**Pull request merged**